src/ride/track.c

```diff
@@ -4921,15 +4922,493 @@ void game_command_remove_track(int *eax, int *ebx, int *ecx, int *edx, int *esi,
+		}
+	}
+
+	// loc_6CD954:
+	if (map_is_location_owned(floor2(x, 32), floor2(y, 32), z) == 0) {
+		// loc_6CDE51
+		return MONEY32_UNDEFINED;
+	}
+
+	rct_map_element *mapElement = map_get_surface_element_at(x / 32, y / 32);
+	if (mapElement == NULL) {
+		// TODO: set error text?
+		return MONEY32_UNDEFINED;
+	}
+
+	// loc_6CD99A:
+	temp_dx = z;
+	temp_dx += 0x20;
+	temp_dx = temp_dx >> 3;
+	sint8 temp_dl = (uint8) (temp_dx & 0xFF);
+
+	temp_dl -= mapElement->base_height;
+	if (temp_dl >= 0) {
+		temp_dl = temp_dl >> 1;
+
+		if (temp_dl > RCT2_GLOBAL(0x0097D218 + (RIDE_TYPE_MAZE * 8), uint8)) {
+			RCT2_GLOBAL(RCT2_ADDRESS_GAME_COMMAND_ERROR_TEXT, rct_string_id) = STR_TOO_HIGH_FOR_SUPPORTS;
+			// loc_6CDE51
+			return MONEY32_UNDEFINED;
+		}
+	}
+
+	// loc_6CD9C5
+	// push    eax
+	// push    ebx
+	// push    ecx
+	// push    edx
+	// push    edi
+
+	// mov     bl, dl
+	uint8 bl = rideIndex;
+	// mov     dx, di
+	temp_dx = z;
+	// shr     dx, 3
+	temp_dx = temp_dx >> 3;
+
+	mapElement = map_get_first_element_at(x / 32, y / 32);
+
+	do {
+		// loc_6CD9F0:
+		if (map_element_get_type(mapElement) != MAP_ELEMENT_TYPE_TRACK) {
+			continue;
+		}
+
+		if ((temp_dx & 0xff) != mapElement->base_height) {
+			continue;
+		}
+
+		if (rideIndex != mapElement->properties.track.ride_index) {
+			continue;
+		}
+		if (mapElement->properties.track.type != 0x65) {
+			continue;
+		}
+
+		return loc_6CDB35(x, flags, direction, y, rideIndex, mode, unknown, z, mapElement);
+	} while (!map_element_is_last_for_tile(mapElement++));
+
+
+	if (mode != 0) {
+		// loc_6CDE43
+		RCT2_GLOBAL(RCT2_ADDRESS_GAME_COMMAND_ERROR_TEXT, unsigned short) = 0;
+		// loc_6CDE4C
+		// loc_6CDE51
+		return MONEY32_UNDEFINED;
+	}
+
+	if (!map_can_construct_at((uint16) floor2(x, 32), (uint16) floor2(y, 32), (z / 8), (z / 8) + 4, 0x0F)) {
+		// loc_6CDE4C
+		// loc_6CDE51
+		return MONEY32_UNDEFINED;
+	}
+
+	if (RCT2_GLOBAL(RCT2_ADDRESS_ELEMENT_LOCATION_COMPARED_TO_GROUND_AND_WATER, uint8) & 0x04) {
+		// loc_6CDE22
+		RCT2_GLOBAL(RCT2_ADDRESS_GAME_COMMAND_ERROR_TEXT, unsigned short) = STR_RIDE_CANT_BUILD_THIS_UNDERWATER;
+		// loc_6CDE4C
+		// loc_6CDE51
+		return MONEY32_UNDEFINED;
+	}
+
+	if (RCT2_GLOBAL(RCT2_ADDRESS_ELEMENT_LOCATION_COMPARED_TO_GROUND_AND_WATER, uint8) & 0x02) {
+		// loc_6CDE17
+		RCT2_GLOBAL(RCT2_ADDRESS_GAME_COMMAND_ERROR_TEXT, unsigned short) = STR_CAN_ONLY_BUILD_THIS_ABOVE_GROUND;
+		// loc_6CDE4C
+		// loc_6CDE51
+		return MONEY32_UNDEFINED;
+	}
+
+	rct_ride *ride = get_ride(rideIndex);
+
+	money32 price = (((RCT2_ADDRESS(0x0097DD78, money16)[ride->type * 2] * RCT2_GLOBAL(0x0099DBC8, money32)) >> 16));
+	RCT2_GLOBAL(0x00F4413E, money32) = price / 2 * 10;
+
+
+	if ((flags & GAME_COMMAND_FLAG_APPLY) == 0) {
+		return loc_6CDDFD();
+	}
+
+
+	// TODO: shouldn't these arguments get shifted?
+	uint16 flooredX = floor2(x, 32);
+	uint16 flooredY = floor2(y, 32);
+
+	mapElement = map_element_insert(flooredX / 32, flooredY / 32, z >> 3, 0xF);
```

Comment:
There is no point flooring something that will be divided in the next statement.